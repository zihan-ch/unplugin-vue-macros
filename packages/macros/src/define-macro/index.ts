import type { Macro } from './types'

export * from './transform'
export * from './utils'

export const defineMacro = (macro: Macro) => macro
