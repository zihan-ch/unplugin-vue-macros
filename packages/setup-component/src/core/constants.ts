export const SETUP_COMPONENT_ID_SUFFIX = '-setup-component-'
export const SETUP_COMPONENT_ID_REGEX = /-setup-component-(\d+).vue$/
export const SETUP_COMPONENT_SUB_MODULE = /-setup-component-(\d+).vue.*/

export const SETUP_COMPONENT_TYPE = 'SetupFC'
