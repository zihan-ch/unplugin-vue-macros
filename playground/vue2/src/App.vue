<script setup lang="ts">
import DefineOptions from './cases/define-options.vue'
import DefineModel from './cases/define-model/parent.vue'
</script>

<template>
  <div>
    <define-options />
    <define-model />
  </div>
</template>
