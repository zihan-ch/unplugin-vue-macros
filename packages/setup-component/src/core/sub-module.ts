import { SETUP_COMPONENT_SUB_MODULE } from './constants'

export const isSubModule = (id: string) => SETUP_COMPONENT_SUB_MODULE.test(id)

export const getMainModule = (subModule: string, root: string) =>
  root + subModule.replace(SETUP_COMPONENT_SUB_MODULE, '')
