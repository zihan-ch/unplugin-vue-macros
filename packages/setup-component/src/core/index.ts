import {
  DEFINE_SETUP_COMPONENT,
  MagicString,
  babelParse,
  getLang,
  getTransformResult,
  isCallOf,
} from '@vue-macros/common'
import { walk } from 'estree-walker'
import { normalizePath } from '@rollup/pluginutils'
import { isFunction } from '@babel/types'
import {
  SETUP_COMPONENT_ID_REGEX,
  SETUP_COMPONENT_ID_SUFFIX,
  SETUP_COMPONENT_TYPE,
} from './constants'
import { isSubModule } from './sub-module'
import type { Node } from '@babel/types'
import type { HmrContext, ModuleNode } from 'vite'

export * from './constants'

// TODO SWC

interface FileContextComponent {
  code: string
  body: string
  node: Node
}

interface FileContext {
  components: FileContextComponent[]
  imports: string[]
}

export type SetupComponentContext = Record<string, FileContext>

export const scanSetupComponent = (code: string, id: string): FileContext => {
  const program = babelParse(code, getLang(id))

  const components: {
    /** defineSetupComponent(...) */
    fn?: Node
    /** component decl */
    decl: Node
  }[] = []
  const imports: FileContext['imports'] = []

  walk(program, {
    enter(node: Node) {
      // defineSetupComponent(...)
      if (isCallOf(node, DEFINE_SETUP_COMPONENT)) {
        components.push({
          fn: node,
          decl: node.arguments[0],
        })
      } else if (
        node.type === 'VariableDeclarator' &&
        node.id.type === 'Identifier' &&
        node.id.typeAnnotation?.type === 'TSTypeAnnotation' &&
        node.id.typeAnnotation.typeAnnotation.type === 'TSTypeReference' &&
        node.id.typeAnnotation.typeAnnotation.typeName.type === 'Identifier' &&
        node.id.typeAnnotation.typeAnnotation.typeName.name ===
          SETUP_COMPONENT_TYPE &&
        node.init
      ) {
        // const comp: SetupFC = ...
        components.push({
          decl: node.init,
        })
      } else if (node.type === 'ImportDeclaration') {
        imports.push(code.slice(node.start!, node.end!))
      }
    },
  })

  const ctxComponents = components.map(({ decl, fn }): FileContextComponent => {
    if (!isFunction(decl))
      throw new SyntaxError(
        `${DEFINE_SETUP_COMPONENT}: invalid setup component definition`
      )

    const body = decl?.body
    let bodyStart = body.start!
    let bodyEnd = body.end!
    if (body.type === 'BlockStatement') {
      bodyStart++
      bodyEnd--
    }

    return {
      code: code.slice(decl.start!, decl.end!),
      body: code.slice(bodyStart, bodyEnd),
      node: fn || decl,
    }
  })

  return {
    components: ctxComponents,
    imports,
  }
}

export const transformSetupComponent = (
  code: string,
  id: string,
  ctx: SetupComponentContext
) => {
  const normalizedId = normalizePath(id)
  const s = new MagicString(code)

  const fileContext = scanSetupComponent(code, id)
  ctx[normalizedId] = fileContext

  for (const [i, { node }] of fileContext.components.entries()) {
    const importName = `setupComponent_${i}`

    s.overwrite(node.start!, node.end!, importName)

    s.prepend(
      `import ${importName} from '${normalizedId}${SETUP_COMPONENT_ID_SUFFIX}${i}.vue'\n`
    )
  }

  return getTransformResult(s, id)
}

export const loadSetupComponent = async (
  virtualId: string,
  ctx: SetupComponentContext,
  root: string
) => {
  const index = +(SETUP_COMPONENT_ID_REGEX.exec(virtualId)?.[1] ?? -1)
  const id = virtualId.replace(SETUP_COMPONENT_ID_REGEX, '')
  const { components, imports } = ctx[id] || ctx[root + id] || {}
  const component = components[index]
  if (!component) return

  const { body } = component
  const lang = getLang(id)

  const s = new MagicString(body)
  const program = babelParse(body, lang, {
    allowReturnOutsideFunction: true,
    allowImportExportEverywhere: true,
  })
  for (const stmt of program.body) {
    // transform return
    if (stmt.type !== 'ReturnStatement' || !stmt.argument) continue
    s.overwriteNode(stmt, `defineRender(${s.sliceNode(stmt.argument)});`)
  }

  for (const i of imports) s.prepend(`${i}\n`)

  s.prepend(`<script setup${lang ? ` lang="${lang}"` : ''}>`)
  s.append(`</script>`)

  return s.toString()
}

export const hotUpdateSetupComponent = async (
  { file, modules, read }: HmrContext,
  ctx: SetupComponentContext
) => {
  const getSubModule = (module: ModuleNode): ModuleNode[] => {
    const importedModules = Array.from(module.importedModules)
    if (importedModules.length === 0) return []

    return importedModules
      .filter(({ id }) => id && isSubModule(id!))
      .flatMap((module) => [module, ...getSubModule(module)])
  }

  const module = modules.find((mod) => mod.file === file)
  if (!module?.id) return

  const affectedModules = getSubModule(module)

  const normalizedId = normalizePath(file)
  const nodeContexts = scanSetupComponent(await read(), normalizedId)
  ctx[normalizedId] = nodeContexts

  return [...modules, ...affectedModules]
}
