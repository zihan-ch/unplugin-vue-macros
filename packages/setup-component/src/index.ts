import { createUnplugin } from 'unplugin'
import { createFilter } from '@rollup/pluginutils'
import { REGEX_SETUP_SFC, REGEX_SRC_FILE } from '@vue-macros/common'
import {
  SETUP_COMPONENT_ID_REGEX,
  hotUpdateSetupComponent,
  loadSetupComponent,
  transformSetupComponent,
} from './core'
import { getMainModule, isSubModule } from './core/sub-module'
import type { PluginContext } from 'rollup'
import type { FilterPattern } from '@rollup/pluginutils'
import type { SetupComponentContext } from './core'

export interface Options {
  include?: FilterPattern
  exclude?: FilterPattern
  root?: string
}

export type OptionsResolved = Omit<Required<Options>, 'exclude'> & {
  exclude?: FilterPattern
}

function resolveOption(options: Options): OptionsResolved {
  return {
    include: [REGEX_SRC_FILE],
    exclude: [REGEX_SETUP_SFC],
    root: process.cwd(),
    ...options,
  }
}

const name = 'unplugin-vue-setup-component'

export default createUnplugin((userOptions: Options = {}, meta) => {
  const options = resolveOption(userOptions)
  const filter = createFilter(options.include, options.exclude)

  const setupComponentContext: SetupComponentContext = {}

  return {
    name,
    enforce: 'pre',

    resolveId(id, importer) {
      if (SETUP_COMPONENT_ID_REGEX.test(id)) return id

      if (
        ['rollup', 'vite'].includes(meta.framework) &&
        importer &&
        isSubModule(importer)
      ) {
        const mainModule = getMainModule(importer, options.root)
        return (this as unknown as PluginContext).resolve(id, mainModule, {
          skipSelf: true,
        }) as any
      }
    },

    loadInclude(id) {
      return SETUP_COMPONENT_ID_REGEX.test(id)
    },

    load(id) {
      return loadSetupComponent(id, setupComponentContext, options.root)
    },

    transformInclude(id) {
      return filter(id)
    },

    transform(code, id) {
      try {
        return transformSetupComponent(code, id, setupComponentContext)
      } catch (err: unknown) {
        this.error(`${name} ${err}`)
      }
    },

    vite: {
      configResolved(config) {
        options.root = config.root
      },

      handleHotUpdate: (ctx) => {
        if (filter(ctx.file)) {
          return hotUpdateSetupComponent(ctx, setupComponentContext)
        }
      },
    },
  }
})
