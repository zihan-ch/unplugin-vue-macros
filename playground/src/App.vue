<script setup lang="ts">
import { ref } from 'vue'

import DefineRenderVue from './demo/define-render.vue'
import DefineOptionsVue from './demo/define-options.vue'
import DefineOptionsEmptyVue from './demo/define-options-empty.vue'
import DefineOptionsTsx from './demo/define-options-tsx.vue'
import HoistStaticVue from './demo/hoist-static.vue'
import DefineModelVue from './demo/define-model.vue'
import { SetupComponentFoo } from './demo/setup-component'
import SetupSFC from './demo/setup-sfc.setup'
import Full from './demo/full.setup'
import { SetupFCType } from './demo/setup-fc.type'

const title = ref('hello')
</script>

<template>
  <DefineRenderVue />
  <hr />
  <DefineOptionsVue />
  <hr />
  <DefineOptionsEmptyVue />
  <hr />
  <DefineOptionsTsx />
  <hr />
  <HoistStaticVue />
  <hr />
  <DefineModelVue v-model:title="title" />
  <hr />
  <SetupComponentFoo />
  <hr />
  <SetupSFC />
  <hr />
  <Full />
  <hr />
  <SetupFCType />
</template>
