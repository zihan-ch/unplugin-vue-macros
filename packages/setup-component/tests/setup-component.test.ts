import { resolve } from 'node:path'
import { describe, expect, it, test } from 'vitest'
import glob from 'fast-glob'
import {
  RollupEsbuildPlugin,
  RollupRemoveVueFilePathPlugin,
  RollupVue,
  RollupVueJsx,
  rollupBuild,
} from '@vue-macros/test-utils'
import { babelParse } from '@vue-macros/common'
import VueSetupComponent from '../src/rollup'
import { overwriteImportsPath } from '../src/core'

describe('setup-component', () => {
  describe('fixtutes', async () => {
    const root = resolve(__dirname, '..')
    const files = await glob('tests/fixtures/*.{vue,[jt]s?(x)}', {
      cwd: root,
      onlyFiles: true,
    })

    for (const file of files) {
      it(file.replace(/\\/g, '/'), async () => {
        const filepath = resolve(root, file)

        const code = await rollupBuild(filepath, [
          VueSetupComponent(),
          RollupVue(),
          RollupVueJsx(),
          RollupRemoveVueFilePathPlugin(),
          RollupEsbuildPlugin({
            target: 'esnext',
          }),
        ])
        expect(code).toMatchSnapshot()
      })
    }
  })

  test('overwriteImportsPath', () => {
    function testImport(code: string) {
      expect(
        overwriteImportsPath(code, babelParse(code).body as any[]).join('\n')
      ).matchSnapshot()
    }

    testImport(`import { a, b as c } from "./d";
import { a1, b2 as b3 } from "../d"`)
    testImport('import { a, b as c } from "ref"')
  })
})
