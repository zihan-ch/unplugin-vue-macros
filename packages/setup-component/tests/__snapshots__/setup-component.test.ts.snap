// Vitest Snapshot v1

exports[`setup-component > tests/fixtures/SetupFC.tsx 1`] = `
"import { defineComponent, ref, createVNode, createTextVNode } from 'vue';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"2\\",
  setup(__props) {
    const count = ref(0);
    defineRender(() => createVNode(\\"div\\", null, [createVNode(\\"p\\", null, [createTextVNode(\\"hi, this is \\"), props.name]), createVNode(\\"p\\", null, [count.value]), createVNode(\\"button\\", {
      \\"onClick\\": () => count.value++
    }, [createTextVNode(\\"inc\\")])]));
    return () => {
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var setupComponent_2 = /*#__PURE__*/_export_sfc(_sfc_main, [__FILE__]);

const App = setupComponent_2;

export { App };
"
`;

exports[`setup-component > tests/fixtures/async.tsx 1`] = `
"import { defineComponent, withAsyncContext } from 'vue';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"0\\",
  props: {
    foo: {
      type: String,
      required: true
    }
  },
  async setup(__props) {
    let __temp, __restore;
    [__temp, __restore] = withAsyncContext(() => globalThis.foo()), await __temp, __restore();
    return () => {
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var setupComponent_0 = /*#__PURE__*/_export_sfc(_sfc_main, [__FILE__]);

const App = setupComponent_0;

export { App };
"
`;

exports[`setup-component > tests/fixtures/basic.ts 1`] = `
"import { defineComponent, h } from 'vue';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"1\\",
  props: {
    foo: { type: String, required: true }
  },
  emits: [\\"change\\"],
  setup(__props) {
    defineOptions({
      name: \\"App\\"
    });
    return () => {
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var setupComponent_1 = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

const App = setupComponent_1;
h(App);

export { App };
"
`;

exports[`setup-component > tests/fixtures/render.tsx 1`] = `
"import { defineComponent, createVNode } from 'vue';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"0\\",
  props: {
    foo: {
      type: String,
      required: true
    }
  },
  setup(__props) {
    defineRender(() => createVNode(\\"div\\", null, null));
    return () => {
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var setupComponent_0 = /*#__PURE__*/_export_sfc(_sfc_main, [__FILE__]);

const App = setupComponent_0;

export { App };
"
`;

exports[`setup-component > tests/fixtures/return-render.tsx 1`] = `
"import { defineComponent, computed, createVNode } from 'vue';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"1\\",
  props: {
    foo: {
      type: String,
      required: true
    }
  },
  setup(__props) {
    const sum = computed(() => 1 + 2);
    defineRender(() => createVNode(\\"div\\", null, [sum.value]));
    return () => {
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var setupComponent_1 = /*#__PURE__*/_export_sfc(_sfc_main, [__FILE__]);

const App = setupComponent_1;

export { App };
"
`;

exports[`setup-component > tests/fixtures/type-annotation.ts 1`] = `
"import { defineComponent, h } from 'vue';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"1\\",
  props: {
    foo: { type: String, required: true }
  },
  emits: [\\"change\\"],
  setup(__props) {
    defineOptions({
      name: \\"App\\"
    });
    return () => {
    };
  }
});

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

var setupComponent_1 = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

const App = setupComponent_1;
h(App);

export { App };
"
`;
